File: Week_07/G20200343040094/51.n皇后.java
import java.awt.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;

/*
 * @lc app=leetcode.cn id=51 lang=java
 *
 * [51] N皇后
 *
 * https://leetcode-cn.com/problems/n-queens/description/
 *
 * algorithms
 * Hard (68.72%)
 * Likes:    384
 * Dislikes: 0
 * Total Accepted:    36.5K
 * Total Submissions: 53K
 * Testcase Example:  '4'
 *
 * n 皇后问题研究的是如何将 n 个皇后放置在 n×n 的棋盘上，并且使皇后彼此之间不能相互攻击。
 * 
 * 
 * 
 * 上图为 8 皇后问题的一种解法。
 * 
 * 给定一个整数 n，返回所有不同的 n 皇后问题的解决方案。
 * 
 * 每一种解法包含一个明确的 n 皇后问题的棋子放置方案，该方案中 'Q' 和 '.' 分别代表了皇后和空位。
 * 
 * 示例:
 * 
 * 输入: 4
 * 输出: [
 * ⁠[".Q..",  // 解法 1
 * ⁠ "...Q",
 * ⁠ "Q...",
 * ⁠ "..Q."],
 * 
 * ⁠["..Q.",  // 解法 2
 * ⁠ "Q...",
 * ⁠ "...Q",
 * ⁠ ".Q.."]
 * ]
 * 解释: 4 皇后问题存在两个不同的解法。
 * 
 * 
 */

// @lc code=start
class Solution {
    Set<Integer> cols = new HashSet<>();
    Set<Integer> diag1 = new HashSet<>();
    Set<Integer> diag2 = new HashSet<>();

    public List<List<String>> solveNQueens(int n) {
        List<List<String>> res = new LinkedList<>();
        dfs(res, new LinkedList<>(), 0, n);
        return res;
    }

    private void dfs(List<List<String>> res, List<String> tempList, int row, int n) {
        if (row == n) {
            res.add(new LinkedList<>(tempList));
            return;
        }
        for (int i = 0; i < n; ++i) {
            if (cols.contains(i) || diag1.contains(row + i) || diag2.contains(row - i)) continue;
            
            char[] charArr = new char[n];
            Arrays.fill(charArr, '.');
            charArr[i] = 'Q';
            String rowStr = String.valueOf(charArr);
            
            tempList.add(rowStr);
            cols.add(i);
            diag1.add(row + i);
            diag2.add(row - i);

            dfs(res, tempList, row + 1, n);

            tempList.remove(tempList.size() - 1);
            cols.remove(i);
            diag1.remove(row + i);
            diag2.remove(row - i);
        }
    }
}
// @lc code=end

File: Week_08/G20200343040262/51.n皇后.cpp
/*
 * @lc app=leetcode.cn id=51 lang=cpp
 *
 * [51] N皇后
 */

// @lc code=start
class Solution {
public:
    //根据qPos(皇后的位置)的值，构造输出的字符串
    void make_ret(int n, const vector<int> &qPos, vector<vector<string>> &ret) {
        vector<string> temp(n, string(n, '.'));
        for (int c = 0; c < n; c++)
            temp[qPos[c]][c] = 'Q';
        ret.push_back(temp);
    }

    void backtrack(int r, int n, int col, int hill, int dale, vector<int> &qPos, vector<vector<string>> &ret) {
        if (r >= n)
            make_ret(n, qPos, ret);
        else
            for (int c = 0; c < n; c++)
                //检查col,hill,dale特定位上的值
                //如(1&col>>c)==0，就是检查col第c个比特位是否==0
                // r表示行下标，c表示列下标，r+c与r-c值均为常数，注意：r-c有可能小于0，所以要+n
                if ((1 & col >> c) == 0 && (1 & hill >> r + c) == 0 && (1 & dale >> r - c + n) == 0) {
                    qPos[c] = r;
                    //将col,hill,dale特定位的0置为1
                    //如col|1<<c，就是将col第c个比特位的0变为1
                    backtrack(r + 1, n, col | 1 << c, hill | 1 << r + c, dale | 1 << r - c + n, qPos, ret);
                    qPos[c] = 0;
                }
    }

    vector<vector<string>> solveNQueens(int n) {
        vector<vector<string>> ret;
        vector<int> qPos(n);
        backtrack(0, n, 0, 0, 0, qPos, ret);
        return ret;
    }
};
// @lc code=end
